tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java

```diff
@@ -534,23 +561,47 @@ public void importGraphDef(GraphDef graphDef, String prefix) throws IllegalArgum
    * @see #importGraphDef(GraphDef, String)
    */
   public GraphDef toGraphDef() {
+    addInitOp();
     synchronized (nativeHandleLock) {
       return toGraphDef(nativeHandle);
     }
   }
 
-  /**
-   * Adds an initializer to the graph initializer list.
-   *
-   * @param initializer An initializer to add to the list.
-   */
-  public synchronized void addInitializer(Op initializer) {
-    initializers.add(initializer);
+  @Override
+  public synchronized void registerInitOp(Operation op) {
+    if (isInitOp(op)) return;
+    checkInput(op);
+
+    if (!(op instanceof GraphOperation)) {
+      throw new IllegalArgumentException("Can't use a non-graph op as a graph's init op.");
+    }
+    GraphOperation graphOp = (GraphOperation) op;
+
+    if (op.type().equals(Placeholder.OP_NAME)) {
+      throw new IllegalArgumentException("Can not make a placeholder an init op.");
+    }
+
+    graphOp.controlInputs().forEach(this::registerInitOp);
+    graphOp.inputs().forEach(x -> registerInitOp(x.op()));
+
+    initializers.removeAll(subgraphToOps(Collections.singleton(graphOp)));
+    initializers.add(op);
+    newInitializers = true;
```

Comment:
What happens with such sequence of calls:
```
graph.registerInitOp(x);
graph.toGraphDef();
graph.registerInitOp(y);
graph.toGraphDef();
```

`toGraphDef()` will add two op `INIT_OP_BASE_NAME`, which should fail right? Is that the expected behaviour?

Reply:
It won't, since the Scope will uniquify it.  I'm pretty sure I test that somewhere.